### PR TITLE
[SDK] Ensure correct UA in SDK/CLI telemetry 

### PR DESCRIPTION
**Pull request by D-W-**
This pull request includes various changes to improve telemetry functionality and testing. The most important changes include adding new test configurations for telemetry, modifying test methods to verify user agent sent by the SDK and Azure CLI, and updating code to include user agent in telemetry logging.

Telemetry improvements:

* <a href="diffhunk://#diff-83ec82bc11bb147db2c60542c1d9c933b4656f6c610fa8f2ab373fed8f7af5ceR1-R146">`src/promptflow/tests/test_configs/recordings/test_telemetry_TestTelemetry_test_custom_event.yaml`</a>: Added a new test configuration file to test telemetry interactions with Azure Management APIs.
* <a href="diffhunk://#diff-b2ec38c5870b7faa23ef68ce3076a71978711f94d9877060dea64451e9bc747fR146-R189">`src/promptflow/tests/sdk_cli_azure_test/e2etests/test_telemetry.py`</a>: Added a new test method to verify the user agent sent by the SDK for telemetry purposes.
* <a href="diffhunk://#diff-d5a4a429f9ab76787a5800c3c679ac45e94ca9b5193e3f726ab9d528397ee339R151-R164">`src/promptflow/tests/sdk_cli_azure_test/e2etests/test_cli_with_azure.py`</a>: Added a new test method to verify the user agent sent by the Azure CLI for telemetry purposes.
* <a href="diffhunk://#diff-e8775c719ab4ac2bb3aef826ea48211abc031f3d56c8a630893fde2c4a577c02R54-R63">`src/promptflow/promptflow/_telemetry/activity.py`</a>: Modified the `log_activity` function to include the user agent in the telemetry information.
* <a href="diffhunk://#diff-3f1dd7459115086f5d3af3d298744079416aebf0a0a99d19883591e7e053cf19L85-R96">`src/promptflow/promptflow/_cli/_pf_azure/entry.py`</a>: Modified the main function to set the user agent in the operation context.

Testing improvements:

* <a href="diffhunk://#diff-7d74ecb42a689e4b7dea2171b91f90db56f79b6efcd70629df1548a1b091f229R22-R26">`src/promptflow/tests/sdk_cli_test/e2etests/test_cli.py`</a>: Added a new test case to check if the user agent dictionary contains the expected keys.
* <a href="diffhunk://#diff-b2ec38c5870b7faa23ef68ce3076a71978711f94d9877060dea64451e9bc747fL6-L17">`src/promptflow/tests/sdk_cli_azure_test/e2etests/test_telemetry.py`</a>: Modified imports in the test file and modified a function to use a different decorator. <a href="diffhunk://#diff-b2ec38c5870b7faa23ef68ce3076a71978711f94d9877060dea64451e9bc747fL6-L17">[1]</a> <a href="diffhunk://#diff-b2ec38c5870b7faa23ef68ce3076a71978711f94d9877060dea64451e9bc747fL48-R50">[2]</a>

Code modifications:

* <a href="diffhunk://#diff-47208ac35b30920275fcd5e55d662647ef360129359bdc77fddd2a2157b6f47eL790-R802">`src/promptflow/promptflow/_sdk/_utils.py`</a>: Made the `user_agent` parameter optional with a default value.
* <a href="diffhunk://#diff-9679bafe4c1b51de85c4389b8b28446f2ab9db89b082c0f1435202d90be24494L13-R12">`src/promptflow/promptflow/_cli/_pf_azure/_utils.py`</a>: Removed the `user_agent` argument from the `PFClient` constructor.
* <a href="diffhunk://#diff-d80d70a0879d410337d9905c070a80049e5fa2808ba4f4089c78625067ae32d7L9-L33">`src/promptflow/promptflow/_telemetry/logging_handler.py`</a>: Removed the `USER_AGENT` constant and the `setup_user_agent_to_operation_context` method.

Configuration changes:

* <a href="diffhunk://#diff-02562d9849b7597305422e9b1dd0ac200c59998feec5522d07824726b40808caR1-R50">`src/promptflow/tests/test_configs/recordings/test_cli_with_azure_TestCliWithAzure_test_azure_cli_ua.yaml`</a>: Added a new interaction to test the behavior of the system when a request with a specific user agent is made.# Description

Please add an informative description that covers that changes made by the pull request and link all relevant issues.

# All Promptflow Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes].**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [ ] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: src/promptflow/promptflow/_telemetry/activity.py [hunk 1]
```diff
@@ -51,11 +51,16 @@ def log_activity(
     :type custom_dimensions: dict
     :return: None
     """
+    from promptflow._sdk._utils import setup_user_agent_to_operation_context
+
+    user_agent = setup_user_agent_to_operation_context()
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_telemetry/activity.py:56]: maybe change to get_xx

File: src/promptflow/promptflow/_utils/utils.py [hunk 2]
```diff
@@ -235,3 +236,14 @@ def resolve_dir_to_absolute(base_dir: Union[str, Path], sub_dir: Union[str, Path
...
     return path
+
+
+def parse_ua_to_dict(ua):
+    """Parse string user agent to dict with name as ua name and value as ua version."""
+    ua_dict = {}
+    ua_list = ua.split(" ")
+    for item in ua_list:
+        if item:
+            key, value = item.split("/")
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_utils/utils.py:247]: what if the item does not have /?

File: src/promptflow/tests/sdk_cli_azure_test/e2etests/test_cli_with_azure.py [hunk 3]
```diff
@@ -146,3 +148,17 @@ def test_run_file_with_set(self, pf: PFClient, runtime: str, randstr: Callable[[
         run = pf.runs.get(run=name)
         assert isinstance(run, Run)
         assert run.properties["azureml.promptflow.runtime_name"] == runtime
+
+    def test_azure_cli_ua(self, pf):
+        with environment_variable_overwrite("USER_AGENT", ""):
```
Comment (wangchao1230) [inline src/promptflow/tests/sdk_cli_azure_test/e2etests/test_cli_with_azure.py:153]: We shall accept environment variable: PF_USER_AGENT

like existing one: 
PF_NO_INTERACTIVE_LOGIN = "PF_NO_INTERACTIVE_LOGIN"

File: src/promptflow/tests/sdk_cli_azure_test/e2etests/test_telemetry.py [hunk 4]
```diff
@@ -142,3 +143,47 @@ def test_cached_logging_handler(self):
...
+            ua_dict = parse_ua_to_dict(ua)
+            assert ua_dict.keys() == {"promptflow-sdk", "promptflow"}
+
+        logger = MagicMock()
+        logger.info = MagicMock()
+        logger.info.side_effect = assert_ua
+
+        # get telemetry logger from SDK should not have extension ua
+        # start a clean local SDK client
+        with environment_variable_overwrite("USER_AGENT", ""):
```
Comment (wangchao1230) [inline src/promptflow/tests/sdk_cli_azure_test/e2etests/test_telemetry.py:163]: make this a constant

File: src/promptflow/tests/sdk_cli_azure_test/e2etests/test_cli_with_azure.py (hunk 3 above)
Comment (D-W-) [reply src/promptflow/tests/sdk_cli_azure_test/e2etests/test_cli_with_azure.py:153]: Do we support both USER_AGENT and PF_USER_AGENT?

Comment (wangchao1230) [reply src/promptflow/tests/sdk_cli_azure_test/e2etests/test_cli_with_azure.py:153]: not sure which part uses the USER_AGENT env, not in my sight

Comment (wangchao1230) [reply src/promptflow/tests/sdk_cli_azure_test/e2etests/test_cli_with_azure.py:153]: maybe it's vscode extension

File: src/promptflow/promptflow/_telemetry/activity.py (hunk 1 above)
Comment (Stephen1993) [reply src/promptflow/promptflow/_telemetry/activity.py:56]: OperationContext.get_instance().get_user_agent()?

File: src/promptflow/promptflow/_sdk/_utils.py [hunk 5]
```diff
@@ -787,14 +787,20 @@ def update_user_agent_from_env_var():
         OperationContext.get_instance().append_user_agent(os.environ["USER_AGENT"])
 
 
-def setup_user_agent_to_operation_context(user_agent):
-    """Setup user agent to OperationContext"""
+def setup_user_agent_to_operation_context(user_agent=None):
```
Comment (Stephen1993) [inline src/promptflow/promptflow/_sdk/_utils.py:790]: Is this function redundant? Didn't we do more logic?
We can directly modify and use the `append_user_agent()`?

File: src/promptflow/promptflow/_telemetry/activity.py (hunk 1 above)
Comment (Stephen1993) [reply src/promptflow/promptflow/_telemetry/activity.py:56]: I suggest we keep only the `append_user_agent` and `get_user_agent `functions. Adding too many functions for modifying the agent might lead to mistake by others use. Therefore, if necessary, should the function `update_user_agent_from_env_var() `be merged into `append_user_agent`?

File: src/promptflow/promptflow/_sdk/_utils.py (hunk 5 above)
Comment (Stephen1993) [reply src/promptflow/promptflow/_sdk/_utils.py:790]: If there must be a set_and_get, Is it more appropriate to place the get function in the OperationContext? The name has been changed to set_and_get_user_agent?

Comment (D-W-) [reply src/promptflow/promptflow/_sdk/_utils.py:790]: This function will add ua from env var to context

File: src/promptflow/promptflow/_utils/utils.py (hunk 2 above)
Comment (D-W-) [reply src/promptflow/promptflow/_utils/utils.py:247]: Seems normalized ua will have it. https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/User-Agent

File: src/promptflow/tests/sdk_cli_azure_test/e2etests/test_cli_with_azure.py (hunk 3 above)
Comment (D-W-) [reply src/promptflow/tests/sdk_cli_azure_test/e2etests/test_cli_with_azure.py:153]: Offline synced, will support both

File: src/promptflow/promptflow/_telemetry/activity.py [hunk 6]
```diff
@@ -51,11 +51,16 @@ def log_activity(
...
+
     activity_info = {
         # TODO(2699383): use same request id with service caller
         "request_id": str(uuid.uuid4()),
         "activity_name": activity_name,
         "activity_type": activity_type,
+        "user_agent": user_agent,
```
Comment (Stephen1993) [inline src/promptflow/promptflow/_telemetry/activity.py:63]: It seems that our current logs are already targeting user_agent, the user_agent added here directly overwrites the previous? Do we need to do any merging?
image

Comment (D-W-) [reply src/promptflow/promptflow/_telemetry/activity.py:63]: We'll de-dup when adding, see `append_user_agent`'s implementation. Also, we only add it when CLI/SDK starts. Most time we'll be reading it.

File: src/promptflow/tests/sdk_cli_test/e2etests/test_cli.py [hunk 7]
```diff
@@ -1537,3 +1541,19 @@ def test_data_scrubbing(self):
...
+            with pytest.raises(SystemExit):
+                run_pf_command(
+                    "run",
+                    "show",
+                    "--name",
+                    "not_exist",
+                )
+        user_agent = setup_user_agent_to_operation_context()
+        ua_dict = parse_ua_to_dict(user_agent)
+        assert ua_dict.keys() == {"promptflow-sdk", "promptflow-cli", "promptflow"}
```
Comment (wangchao1230) [inline src/promptflow/tests/sdk_cli_test/e2etests/test_cli.py:1559]: why we have promptflow-sdk here?

Comment (D-W-) [reply src/promptflow/tests/sdk_cli_test/e2etests/test_cli.py:1559]: This command will go through SDK's logic, it refers to this log:  
image

**Pull request merged**